docs/_advanced/error-handling.md

```diff
@@ -221,12 +222,115 @@ export RUBYLLM_DEBUG=true
+chat.with_failover(
+  { provider: :bedrock, model: 'claude-3-5-haiku-20241022' },
+  'gpt-4o-mini' # String shorthand for OpenAI models
+)
+
+# This will automatically failover if the primary provider hits rate limits
+response = chat.ask "What is the capital of France?"
+```
+
+### Failover Configuration Options
+
+The `with_failover` method accepts multiple configuration formats:
+
+#### Hash Configuration (Full Control)
+```ruby
+chat.with_failover(
+  { provider: :bedrock, model: 'claude-3-5-haiku-20241022' },
+  { provider: :openai, model: 'gpt-4o-mini' }
+)
+```
+
+#### String Configuration (Model Name Only)
+```ruby
+# Uses the default provider for the model
+chat.with_failover('gpt-4o-mini', 'gemini-2.0-flash')
+```
+
+#### Mixed Configuration
+```ruby
+chat.with_failover(
+  { provider: :bedrock, model: 'claude-3-5-haiku-20241022' },
+  'gpt-4o-mini', # String shorthand
+  { provider: :gemini, model: 'gemini-2.0-flash' }
+)
+```
+
+### Failover with Different Contexts
+
+You can use different API keys or configurations for failover providers:
+
+```ruby
+# Create a backup context with different credentials
+backup_context = RubyLLM.context do |config|
+  config.anthropic_api_key = 'backup-api-key'
+end
+
+chat = RubyLLM.chat(provider: :anthropic, model: 'claude-3-5-haiku-20241022')
+chat.with_failover(
+  { provider: :anthropic, model: 'claude-3-5-haiku-20241022', context: backup_context }
+)
+```
+
+### When Failover Occurs
+
+Failover is triggered by:
+
+*   **Rate Limit Errors (`RubyLLM::RateLimitError`)** - The primary use case
+*   **Service Unavailable Errors (`RubyLLM::ServiceUnavailableError`)** - When the provider is temporarily down
+*   **Overloaded Errors (`RubyLLM::OverloadedError`)** - When the provider is overloaded
+*   **Server Errors (`RubyLLM::ServerError`)** - When the provider returns 5xx errors
+
+**Note:** Failover does **not** occur for client errors like `BadRequestError` (400), `UnauthorizedError` (401), or `ForbiddenError` (403), as these typically indicate configuration issues that would affect all providers.
+
+### Failover Behavior
+
+1.  RubyLLM attempts the request with the primary provider/model
+2.  If a failover-eligible error occurs, it tries the first backup configuration
+3.  If that also fails with a failover-eligible error, it tries the next backup
+4.  This continues until either a request succeeds or all configurations are exhausted
+5.  If all configurations fail, the last error is raised
+
+### Example: Production-Ready Setup
+
+```ruby
+# Primary: Fast, cheap model
+# Backup: More reliable but expensive model  
+# Final: Different provider entirely
+chat = RubyLLM.chat(provider: :anthropic, model: 'claude-3-5-haiku-20241022')
+chat.with_failover(
+  { provider: :anthropic, model: 'claude-3-5-sonnet-20241022' }, # More capable model
+  { provider: :openai, model: 'gpt-4o' }, # Different provider
+  { provider: :bedrock, model: 'claude-3-5-haiku-20241022' } # AWS backup
+)
```

Comment:
Just to leave you a quick review: after seeing it I'm actually not a big fan of this syntax.

Please don't jump into implementing it yet, but let's explore the option of having an object to which we pass multiple `RubyLLM.chat` or `context.chat` and it simply handles switching between them? I think that could look a bit better.

What do you think?